File: 0392_is_subsequence/python/solution_followup_test.py
import pytest

from solution_followup import SubsequenceChecker


@pytest.fixture
def checker() -> SubsequenceChecker:
    return SubsequenceChecker("ahbgdc")


@pytest.mark.parametrize(
    "s, expected",
    [
        ("abc", True),
        ("axc", False),
        ("", True),
        ("ahbgdc", True),
        ("a", True),
        ("c", True),
        ("aaa", False),
        ("abdc", True),
        ("ahbgdca", False),
    ],
)
def test_followup_against_ahbgdc(checker: SubsequenceChecker, s: str, expected: bool) -> None:
    assert checker.is_subsequence(s) == expected


def test_empty_t() -> None:
    checker = SubsequenceChecker("")
    assert checker.is_subsequence("") is True
    assert checker.is_subsequence("a") is False


def test_single_char_t() -> None:
    checker = SubsequenceChecker("a")
    assert checker.is_subsequence("a") is True
    assert checker.is_subsequence("b") is False
    assert checker.is_subsequence("") is True


def test_repeated_chars_in_t() -> None:
    checker = SubsequenceChecker("aabbc")
    assert checker.is_subsequence("abc") is True
    assert checker.is_subsequence("aab") is True
    assert checker.is_subsequence("aabb") is True
    assert checker.is_subsequence("aabbc") is True
    assert checker.is_subsequence("aabbcc") is False


def test_many_queries_same_t() -> None:
    checker = SubsequenceChecker("abcdefghijklmnopqrstuvwxyz")
    assert checker.is_subsequence("ace") is True
    assert checker.is_subsequence("az") is True
    assert checker.is_subsequence("za") is False
    assert checker.is_subsequence("abc") is True
    assert checker.is_subsequence("xyz") is True

File: 0392_is_subsequence/kotlin/solution_followup_test.kt
fun main() {
    val checker = SubsequenceChecker("ahbgdc")
    val tests = listOf(
        "abc" to true,
        "axc" to false,
        "" to true,
        "ahbgdc" to true,
        "a" to true,
        "c" to true,
        "aaa" to false,
        "abdc" to true,
        "ahbgdca" to false,
    )

    for ((s, expected) in tests) {
        val result = checker.isSubsequence(s)
        check(result == expected) { "Expected $expected, got $result for '$s'" }
    }

    val emptyChecker = SubsequenceChecker("")
    check(emptyChecker.isSubsequence(""))
    check(!emptyChecker.isSubsequence("a"))

    val singleChecker = SubsequenceChecker("a")
    check(singleChecker.isSubsequence("a"))
    check(!singleChecker.isSubsequence("b"))
    check(singleChecker.isSubsequence(""))

    val repeatedChecker = SubsequenceChecker("aabbc")
    check(repeatedChecker.isSubsequence("abc"))
    check(repeatedChecker.isSubsequence("aab"))
    check(repeatedChecker.isSubsequence("aabb"))
    check(repeatedChecker.isSubsequence("aabbc"))
    check(!repeatedChecker.isSubsequence("aabbcc"))

    val alphabetChecker = SubsequenceChecker("abcdefghijklmnopqrstuvwxyz")
    check(alphabetChecker.isSubsequence("ace"))
    check(alphabetChecker.isSubsequence("az"))
    check(!alphabetChecker.isSubsequence("za"))
    check(alphabetChecker.isSubsequence("abc"))
    check(alphabetChecker.isSubsequence("xyz"))

    println("All Kotlin follow-up tests passed!")
}
